daemon/status.go

```diff
@@ -136,6 +143,149 @@ func (h *getHealthz) Handle(params GetHealthzParams) middleware.Responder {
 }
 
+type getNodes struct {
+	d *Daemon
+	// mutex to protect the clients map against concurrent access
+	lock.RWMutex
+	// clients maps a client ID to a clusterNodesClient
+	clients map[int64]*clusterNodesClient
+}
+
+func NewGetClusterNodesHandler(d *Daemon) GetClusterNodesHandler {
+	return &getNodes{
+		d:       d,
+		clients: map[int64]*clusterNodesClient{},
+	}
+}
+
+// clientGCTimeout is the time for which the clients are kept. After timeout
+// is reached, clients will be cleaned up.
+const clientGCTimeout = 15 * time.Minute
+
+type clusterNodesClient struct {
+	// mutex to protect the client against concurrent access
+	lock.RWMutex
+	lastSync time.Time
+	*models.ClusterNodeStatus
+}
+
+func (c *clusterNodesClient) NodeAdd(newNode node.Node) error {
+	c.Lock()
+	c.NodesAdded = append(c.NodesAdded, newNode.GetModel())
+	c.Unlock()
+	return nil
+}
+
+func (c *clusterNodesClient) NodeUpdate(oldNode, newNode node.Node) error {
+	c.Lock()
+	c.NodesAdded = append(c.NodesAdded, newNode.GetModel())
+	c.NodesRemoved = append(c.NodesRemoved, oldNode.GetModel())
+	c.Unlock()
+	return nil
+}
+
+func (c *clusterNodesClient) NodeDelete(node node.Node) error {
+	c.Lock()
+	// If the node was added/updated and removed before the clusterNodesClient
+	// was aware of it then we can safely remove it from the list of added
+	// nodes and not set it in the list of removed nodes.
+	found := -1
+	for i, added := range c.NodesAdded {
+		if added.Name == node.Fullname() {
+			found = i
+		}
+	}
+	if found != -1 {
+		c.NodesAdded = append(c.NodesAdded[:found], c.NodesAdded[found+1:]...)
+	} else {
+		c.NodesRemoved = append(c.NodesRemoved, node.GetModel())
+	}
+	c.Unlock()
+	return nil
+}
+
+func (c *clusterNodesClient) NodeValidateImplementation(node node.Node) error {
+	// no-op
+	return nil
+}
+
+func (c *clusterNodesClient) NodeConfigurationChanged(config datapath.LocalNodeConfiguration) error {
+	// no-op
+	return nil
+}
+
+func (h *getNodes) cleanupClients() {
+	past := time.Now().Add(-clientGCTimeout)
+	for k, v := range h.clients {
+		if v.lastSync.Before(past) {
+			delete(h.clients, k)
+		}
+	}
+}
+
+func (h *getNodes) Handle(params GetClusterNodesParams) middleware.Responder {
+	var cns *models.ClusterNodeStatus
+	// If ClientID is not set then we send all nodes, otherwise we will store
+	// the client ID in the list of clients and we subscribe this new client
+	// to the list of clients.
+	if params.ClientID == nil {
+		ns := h.d.getNodeStatus()
+		cns = &models.ClusterNodeStatus{
+			Self:       ns.Self,
+			NodesAdded: ns.Nodes,
+		}
+		return NewGetClusterNodesOK().WithPayload(cns)
+	}
+
+	h.Lock()
+	defer h.Unlock()
+
+	var clientID int64
+	c, exists := h.clients[*params.ClientID]
+	if exists {
+		clientID = *params.ClientID
+	} else {
+		clientID = rand.Int63()
+		// make sure we haven't allocated an existing client ID
+		// if we have then we will return clientID 0.
+		_, exists := h.clients[clientID]
+		if exists {
```

Comment:
We should probably also follow this path in the `clientID == 0` case, ie if 0 was randomly generated?

Reply:
Nice call.